File: lib/src/main/java/me/ameriod/lib/mvp/view/deligate/ActivityDelegateImpl.java
package me.ameriod.lib.mvp.view.deligate;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import me.ameriod.lib.mvp.Mvp;

public class ActivityDelegateImpl <V extends Mvp.View, P extends Mvp.Presenter<V>> implements ActivityDelegate<V, P>{

    @NonNull
    private ActivityDelegateCallback<V, P> delegateCallback;
    private P presenter;

    public ActivityDelegateImpl(@NonNull ActivityDelegateCallback<V, P> delegateCallback) {
        this.delegateCallback = delegateCallback;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onCreate(@Nullable Bundle savedInstanceState) {
        if (presenter == null) {
            presenter = delegateCallback.createPresenter();
        }
        presenter.attachView((V) this);
        if (savedInstanceState != null) {
            presenter.restoreState(savedInstanceState);
        }
    }

    @Override
    public void onSaveInstanceState(@Nullable Bundle savedInstanceState) {
        if(savedInstanceState != null) {
            presenter.saveState(savedInstanceState);
        }
    }

    @Override
    public void onDestroy() {
        presenter.detachView();
    }

    @Override
    public P getPresenter() {
        return presenter;
    }
}

File: lib/src/main/java/me/ameriod/lib/mvp/deligate/ActivityDelegateImpl.kt
package me.ameriod.lib.mvp.deligate

import android.os.Bundle

import me.ameriod.lib.mvp.Mvp

class ActivityDelegateImpl<V : Mvp.View, out P : Mvp.Presenter<V>>(private val delegateCallback:
                                                                   ActivityDelegateCallback<V, P>) : ActivityDelegate<V, P> {
    private var presenter: P? = null

    @Suppress("UNCHECKED_CAST")
    override fun onCreate(savedInstanceState: Bundle?) {
        if (presenter == null) {
            presenter = delegateCallback.createPresenter()
        }
        presenter!!.attachView(delegateCallback.getMvpView())
        if (savedInstanceState != null) {
            presenter!!.restoreState(savedInstanceState)
        }
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        if (outState != null) {
            presenter!!.saveState(outState)
        }
    }

    override fun onDestroy() {
        presenter!!.detachView()
    }

    override fun getPresenter(): P {
        return presenter!!
    }
}
